apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PriorityHeap.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Google.Cloud.Spanner.V1
+{
+    /// <summary>
+    /// A simple sorted list that allows items to return equality for IComparable.CompareTo
+    /// </summary>
+    internal class PriorityHeap<T> where T : IPriorityHeapItem<T>
```

Comment:
Correct me if I'm wrong, but this seems to be more of a sorted list with a binary search algorithm. If that case, you can use `List<T>.BinarySearch` rather than re-implementing it here. The nice thing about this method is the result of a search will be negative if the item is not found, but not just any negative number - it will be the bitwise complement of the index where the item should be inserted if you want to maintain the sort order. So you can implement and add operation like this:
```csharp
int index = _innerList.BinarySearch(item);
if (index < 0)
{
    _innerList.Insert(~index, item);
}
```

A similar simplification can be made for removals.

Also, this is more of a nit, but I wouldn't call it a heap since heaps do O(log N) or better for inserts, deletes, or key changes, but this has O(N) for all of these operations (`List<T>.Insert` and `List<T>.RemoveAt` are O(N) operations).